### PR TITLE
[Feature] Max pool Transform

### PR DESCRIPTION
**Pull request by albertbou92**
## Description

A new Transform that does a max pooling of values over time dimension. It was used, for examples, in https://www.nature.com/articles/nature14236 to deal with flickering objects in Atari 2600 environment.

The PR also includes tests for the Transform.

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: torchrl/envs/transforms/transforms.py [hunk 1]
```diff
@@ -2761,3 +2759,79 @@ def transform_observation_spec(self, observation_spec: TensorSpec) -> TensorSpec
...
         )
+
+
+class TimeMaxPool(Transform):
+    """Take the maximum value in each position over the last T observations.
+
+    This transform take the maximum value in each position for all in_keys tensors over the last T time steps.
+
+    Args:
+        T: int
```
Comment (vmoens) [inline torchrl/envs/transforms/transforms.py:2770]: I'd write
```python
    Args:
        in_keys (sequence of str, optional): input keys on which the max pool will be applied. Defaults to "observation" if left empty.
        out_keys (sequence of str, optional): output keys where the output will be written. Defaults to `in_keys` if left empty.
        T (int, optional): Number of time steps over which to apply max pooling.

```

File: torchrl/envs/transforms/transforms.py [hunk 2]
```diff
@@ -2761,3 +2759,79 @@ def transform_observation_spec(self, observation_spec: TensorSpec) -> TensorSpec
...
+    def __init__(
+        self,
+        in_keys: Optional[Sequence[str]] = None,
+        out_keys: Optional[Sequence[str]] = None,
+        T: int = 1,
+    ):
+        super().__init__(in_keys=in_keys, out_keys=out_keys)
+        if T < 1:
+            raise ValueError(
+                "TimeMaxPoolTranform N parameter should have a value greater or equal to one."
```
Comment (vmoens) [inline torchrl/envs/transforms/transforms.py:2786]: `T` parameter I guess

File: torchrl/envs/transforms/transforms.py [hunk 3]
```diff
@@ -2761,3 +2759,79 @@ def transform_observation_spec(self, observation_spec: TensorSpec) -> TensorSpec
...
+
+    def _call(self, tensordict: TensorDictBase) -> TensorDictBase:
+        """Update the episode tensordict with max pooled keys."""
+        for in_key in self.in_keys:
+            # Lazy init of buffers
+            if in_key not in self._buffers.keys():
+                self._buffers[in_key] = torch.zeros(
+                    self.buffer_size,
+                    *tensordict[in_key].shape,
+                    dtype=tensordict[in_key].dtype,
```
Comment (vmoens) [inline torchrl/envs/transforms/transforms.py:2824]: what about device?

File: torchrl/envs/transforms/transforms.py [hunk 4]
```diff
@@ -2761,3 +2759,79 @@ def transform_observation_spec(self, observation_spec: TensorSpec) -> TensorSpec
                 if key in self.selected_keys
             }
         )
+
+
+class TimeMaxPool(Transform):
```
Comment (vmoens) [inline torchrl/envs/transforms/transforms.py:2764]: I think that in the case where out_keys is not in_keys we'll have a problem with the specs.
Can you address this? We just need to copy the spec of all the in_keys on the out_keys

**Review by vmoens — APPROVED** [review]
Great implementation i like it
Can you address the small comments i left?

**Comment by vmoens** [comment]
FYI 
you may want to merge main, we have deprecated SavedTensorDict and tests fail bc of it

**Comment by vmoens** [comment]
I love the logic so much that I will duplicate it for cat frames :p

File: torchrl/envs/transforms/transforms.py [hunk 5]
```diff
@@ -2761,3 +2759,100 @@ def transform_observation_spec(self, observation_spec: TensorSpec) -> TensorSpec
...
+            # add new obs
+            self._buffers[in_key][0].copy_(tensordict[in_key])
+            # apply max pooling
+            pooled_tensor, _ = self._buffers[in_key].max(dim=0)
+            # add to tensordict
+            tensordict.set(out_key, pooled_tensor)
+
+        return tensordict
+
+    def transform_observation_spec(self, observation_spec: TensorSpec) -> TensorSpec:
```
Comment (vmoens) [inline torchrl/envs/transforms/transforms.py:2844]: You should be able to use `@_apply_to_composite` in this case

**Comment by albertbou92** [comment]
I think the registration of buffers should be fine now

**Comment by vmoens** [comment]
I'm #847 I introduced a bug, since uninitialised buffers were not found by load_state_dict.
I solved it in #855.
Sorry about all this
Have a look!

**Comment by albertbou92** [comment]
Oh interesting I did not know about UninitializedBuffer's in PyTorch. I will adapt the code, no problem

**Review by vmoens — APPROVED** [review]
LGTM

**Pull request merged**